src/strategies/offer_maker/market_making/Kandel/abstract/CoreKandel.sol

```diff
@@ -0,0 +1,277 @@
+  {
+    if (msg.value > 0) {
+      MGV.fund{value: msg.value}();
+    }
+    for (uint index = from; index < to; index++) {
+      OfferArgs memory args;
+      OrderType ba = index <= lastBidIndex ? OrderType.Bid : OrderType.Ask;
+      (args.outbound_tkn, args.inbound_tkn) = _tokenPairOfOrderType(ba);
+      args.gives = _givesOfIndex(ba, index);
+      args.wants = _wantsOfIndex(ba, index);
+      args.fund = 0;
+      args.noRevert = false;
+      args.gasreq = offerGasreq();
+      args.gasprice = gasprice;
+      args.pivotId = pivotIds[index - from];
+      _populateIndex(ba, index, args);
+    }
+  }
+
+  ///@notice retracts and deprovisions offers of the distribution interval `[from, to[`
+  ///@param from the start index
+  ///@param to the end index
+  ///@dev this simply provisions this contract's balance on Mangrove.
+  ///@dev use in conjunction of `withdrawFromMangrove` if the user wishes to redeem the available WEIs
+  function retractOffers(uint from, uint to) external onlyAdmin {
+    uint collected;
+    for (uint index = from; index < to; index++) {
+      (IERC20 outbound_tkn, IERC20 inbound_tkn) = _tokenPairOfOrderType(OrderType.Ask);
+      collected +=
+        MGV.retractOffer(address(outbound_tkn), address(inbound_tkn), offerIdOfIndex(OrderType.Ask, index), true);
+      (outbound_tkn, inbound_tkn) = _tokenPairOfOrderType(OrderType.Bid);
+      collected +=
+        MGV.retractOffer(address(outbound_tkn), address(inbound_tkn), offerIdOfIndex(OrderType.Bid, index), true);
+    }
+  }
+
+  ///@notice takes care of reposting residual offer in case of a partial fill and logging potential issues.
+  ///@param order a recap of the taker order
+  ///@param makerData generated during `makerExecute` so as to log it if necessary
+  ///@return notPublished the amount of liquidity that failed to be published on mangrove
+  function _handleResidual(MgvLib.SingleOrder calldata order, bytes32 makerData) internal returns (uint notPublished) {
+    bytes32 repostStatus = super.__posthookSuccess__(order, makerData);
+    // Offer failed to repost for bad reason, logging the incident
+    if (repostStatus == "posthook/filled" || repostStatus == REPOST_SUCCESS) {
+      return 0;
+    }
+    if (repostStatus == "mgv/writeOffer/density/tooLow") {
+      return __residualGives__(order);
+    } else {
+      emit LogIncident(
+        MGV, IERC20(order.outbound_tkn), IERC20(order.inbound_tkn), order.offerId, makerData, repostStatus
+        );
+      return __residualGives__(order);
+    }
+  }
+
+  ///@notice repost residual offer and dual offer according to transport logic
+  ///@inheritdoc MangroveOffer
+  function __posthookSuccess__(MgvLib.SingleOrder calldata order, bytes32 makerData)
+    internal
+    override
+    returns (bytes32)
+  {
+    OrderType ba = _orderTypeOfOutbound(IERC20(order.outbound_tkn));
+    // adds any unpublished liquidity to pending[Base/Quote]
+    setPending(ba, _handleResidual(order, makerData));
```

Comment:
setPending should either increase/decrease or we should add the current pending to the residual? Currently the existing pending is overwritten.